lib/galaxy/tools/parameters/meta.py

```diff
@@ -72,26 +72,30 @@ def expand_workflow_inputs(inputs):
     return params, params_keys
 
 
-def process_key(incoming_key, d):
+def process_key(incoming_key, incoming_value, d):
     key_parts = incoming_key.split('|')
     if len(key_parts) == 1:
         # Regular parameter
-        d[incoming_key] = object()
+        d[incoming_key] = incoming_value
     elif key_parts[0].rsplit('_', 1)[-1].isdigit():
         # Repeat
         input_name_index = key_parts[0].rsplit('_', 1)
         input_name, index = input_name_index
+        index = int(index)
         if input_name not in d:
             d[input_name] = []
-        subdict = {}
-        d[input_name].append(subdict)
-        process_key("|".join(key_parts[1:]), d=subdict)
+        if len(d[input_name]) > index:
+            subdict = d[input_name][index]
+        else:
+            subdict = {}
+            d[input_name].append(subdict)
```

Comment:
@mvdbeek I'm not completely convinced this code is correct: if the repeat elements are not sorted in the `incoming` dictionary, the element with `index` N can be appended here in a position < N.